src/Core/src/MemoryStream.cpp

```diff
@@ -34,25 +58,64 @@ namespace OpenLoco
 
     void MemoryStream::setPosition(size_t position)
     {
-        _offset = std::min(_data.size(), position);
+        _offset = std::min(_length, position);
     }
 
     void MemoryStream::read(void* buffer, size_t len)
     {
-        auto maxReadLen = _data.size() - _offset;
-        if (len > maxReadLen)
+        const auto maxReadLength = std::min(len, _length - _offset);
+        if (len > maxReadLength)
+        {
             throw std::runtime_error("Failed to read data");
-        std::memcpy(buffer, _data.data() + _offset, len);
-        _offset += len;
+        }
+
+        std::memcpy(buffer, _data + _offset, len);
+        _offset += maxReadLength;
+    }
+
+    // TODO: Move this somewhere more sensible.
+    template<typename T>
+    static constexpr T alignTo(T val, T align) noexcept
+    {
+        if (val % align == 0)
+        {
+            return val;
+        }
+        return val + (align - (val % align));
     }
 
     void MemoryStream::write(const void* buffer, size_t len)
     {
-        if (len != 0)
+        const auto spaceLeft = _capacity - _offset;
+
+        // Check if we have to expand the current buffer.
+        if (len > spaceLeft)
         {
-            ensureLength(_offset + len);
-            std::memcpy(_data.data() + _offset, buffer, len);
-            _offset += len;
+            constexpr auto growthFactor = 2.0f;
+            constexpr auto pageSize = 0x1000U;
```

Comment:
kPageSize, kGrowthFactor, ...

Reply:
A bit early to review but yeah, I will probably move this elsewhere and also obtain the correct page size reported by the OS.